### PR TITLE
Prepare 0.24 for rolling upgrade

### LINKED ISSUE #5104: Prepare rolling upgrade from 0.24 to 0.25
**Description**

Issue #5038 changed the serialization format to allow us to make future changes without breaking backwards compatibility. The upgrade path from 0.24 to 0.25 currently requires users to take the cluster offline to  but not rolling upgrades. This is because 0.24 brokers can't read 0.25 messages so new brokers can't join the cluster. To solve this we can have a 0.24 patch that allows "old" brokers to understand both formats but still serialize in the "old" unversioned one. This should allow rolling upgrades from 0.24.x to 0.24.x+1 and then from 0.24.x+1 to 0.25.

### PR DESCRIPTION
**Pull request by miguelpires**
## Description

Allows 0.24 brokers to deserialize messages sent from 0.25 brokers, making it possible for a 0.24 cluster to be rolled to 0.25. I've ran the rolling update test with the 0.25 changes and it works.

## Related issues

closes #5104

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: atomix/utils/src/main/java/io/atomix/utils/serializer/Namespace.java [hunk 1]
```diff
@@ -260,388 +64,16 @@ public void serialize(final Object obj, final OutputStream stream, final int buf
...
-
-    @Override
-    public String toString() {
-      return MoreObjects.toStringHelper(getClass())
-          .add("begin", begin)
-          .add("types", types)
-          .toString();
-    }
+  static NamespaceImpl.Builder builder() {
+    return new NamespaceImpl.Builder();
```
Comment (deepthidevaki) [inline atomix/utils/src/main/java/io/atomix/utils/serializer/Namespace.java:77]: It would be good not to tie implementation with the interface. So I would move this method out and put it in the `NamespaceImpl`.

File: atomix/utils/src/main/java/io/atomix/utils/serializer/FallbackNamespace.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  /**
+   * Deserializes given byte array to Object using Kryo instance in pool.
+   *
+   * @param bytes serialized bytes
+   * @param <T> deserialized Object type
+   * @return deserialized Object
+   */
+  public <T> T deserialize(final byte[] bytes) {
+    try {
+      return namespace.deserialize(bytes);
```
Comment (deepthidevaki) [inline atomix/utils/src/main/java/io/atomix/utils/serializer/FallbackNamespace.java:87]: A broker running this version always serialize with `fallback`, but deserialize with the new one. Is that ideal? A cluster running with all 0.24.x broker then always have to fallback when deserializing. Not sure how much performance impact it will have.

Comment (deepthidevaki) [reply atomix/utils/src/main/java/io/atomix/utils/serializer/FallbackNamespace.java:87]: I think that namespace should be the current one and fallback should be the new one because we have to use fallback only during rolling upgrade to 0.25.x.

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java [hunk 3]
```diff
@@ -147,7 +147,7 @@ private void createSnapshot(final long index) {
     try {
       return builder
           // hardcode max segment size to allow a single entry only
-          .withMaxSegmentSize(JournalSegmentDescriptor.BYTES + 8 * Integer.BYTES)
+          .withMaxSegmentSize(JournalSegmentDescriptor.BYTES + 9 * Integer.BYTES)
```
Comment (deepthidevaki) [inline broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java:150]: Why is this required?

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
Looks good @MiguelPires . Just some small comments.

File: atomix/utils/src/main/java/io/atomix/utils/serializer/FallbackNamespace.java (hunk 2 above)
Comment (MiguelPires) [reply atomix/utils/src/main/java/io/atomix/utils/serializer/FallbackNamespace.java:87]: That was the original idea but some objects serialized with the CompatibleSerializer didn't fail when being deserialized with the FieldSerializer, instead being deserialized to a bad object which caused NPEs. So I had to switch the order. I realize it's not the ideal path but at least I didn't see a meaningful impact in the benchmark I ran

Comment (deepthidevaki) [reply atomix/utils/src/main/java/io/atomix/utils/serializer/FallbackNamespace.java:87]: Ok :+1:

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java (hunk 3 above)
Comment (MiguelPires) [reply broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java:150]: That was necessary in the 0.25 changes because the serialized entry was bigger with the new serializer but here we're still using the old one so it shouldn't be necessary. I just forgot to remove it

**Comment by MiguelPires** [comment]
Thanks for the feedback @deepthidevaki, I addressed your comments in the latest commit

**Review by deepthidevaki — APPROVED** [review]
:+1:

**Comment by deepthidevaki** [comment]
We have reverted this merge from stable/0.24 due to its performance impact. See https://github.com/zeebe-io/zeebe/issues/5300 . We should revisit and see how to merge it again to stable/0.24 before 0.25 release. /cc @MiguelPires
